flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/resource/UserResourceManager.java

```diff
@@ -0,0 +1,195 @@
+            // valid it in here
+            checkJarFile(localUrl);
+
+            // add it to classloader
+            userClassLoader.addURL(localUrl);
+            LOG.info("Added jar resource [{}] to class path.", localUrl.toString());
+        }
+
+        resourceInfos.put(resourceUri, localUrl);
+        LOG.info("Register resource [{}] successfully.", resourceUri.getUri());
+    }
+
+    public Map<ResourceUri, URL> getResources() {
+        return Collections.unmodifiableMap(resourceInfos);
+    }
+
+    public Set<URL> getJarResourceURLs() {
+        return resourceInfos.entrySet().stream()
+                .filter(entry -> ResourceType.JAR.equals(entry.getKey().getResourceType()))
+                .map(Map.Entry::getValue)
+                .collect(Collectors.toSet());
+    }
+
+    private void checkResource(ResourceType resourceType, Path path) {
+        try {
+            FileSystem fs = path.getFileSystem();
+            // check resource exists firstly
+            if (!fs.exists(path)) {
+                throw new IllegalArgumentException(String.format("Resource [%s] not found.", path));
+            }
+
+            if (ResourceType.JAR.equals(resourceType)) {
+                // register directory is not allowed for jar resource
+                if (fs.getFileStatus(path).isDir()) {
+                    throw new IllegalArgumentException(
+                            String.format(
+                                    "Directory [%s] is not allowed for registering jar resource.",
+                                    path));
+                }
+
+                // Jar resource must ends with '.jar' suffix
+                if (!path.getName().toLowerCase().endsWith(".jar")) {
+                    throw new IllegalArgumentException(
+                            String.format(
+                                    "Registering jar resource [%s] must ends with '.jar' suffix.",
+                                    path));
+                }
+            }
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(
+                    String.format("Failed to check resource [%s].", path), e);
+        }
+    }
+
+    @VisibleForTesting
+    protected URL downloadResource(Path remotePath) {
+        // get local resource path
+        java.nio.file.Path path = Paths.get(localTmpDir, remotePath.getName()).toAbsolutePath();
+        Path localResourcePath = new Path(path.toUri());
+        try {
+            FileUtils.copy(remotePath, localResourcePath, true);
+            LOG.info(
+                    "Download resource [{}] to local path [{}] successfully.",
+                    remotePath,
+                    localResourcePath);
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(
+                    String.format(
+                            "Failed to download resource [%s] to local path [%s].",
+                            remotePath, localTmpDir),
```

Comment:
localTmpDir -> localResourcePath? Would be better to print the local file path instead of the base dir?